amazon/ion/writer_binary_raw.py

```diff
@@ -279,9 +287,13 @@ def _serialize_timestamp(ion_event):
                 raise ValueError('Error writing event %s. Found timestamp fractional precision of %d digits, '
                                  'which is less than needed to serialize %d microseconds.'
                                  % (ion_event, fractional_precision, dt.microsecond))
-            exponent = -adjusted_fractional_precision
-            if not (coefficient == 0 and exponent >= 0):
-                length += _write_decimal_value(value_buf, exponent, coefficient)
+            if coefficient_fraction_seconds is None:
+                exponent = -adjusted_fractional_precision
+                if not (coefficient == 0 and exponent >= 0):
+                    length += _write_decimal_value(value_buf, exponent, coefficient)
+
+        if coefficient_fraction_seconds is not None:
```

Comment:
Unfortunately, some unit tests rely on line 287 being thrown, when fractional_precision is less than needed to serialize microseconds. Even if `coefficient_fraction_seconds` is present, unit tests still fail because they expect ValueError to be raised so this check needs to be done. This is why I ordered things this way, to work around this. 
Ideally, as a side task, this check should be moved to the Timestamp constructor and unit tests for writer_binary_raw should be updated accordingly.

Reply:
Okay, this is definitely one of the things that we should call out for cleanup after this PR is merged.

> Even if `coefficient_fraction_seconds` is present, unit tests still fail because they expect ValueError to be raised so this check needs to be done.

We will need to update these tests that expect failure such that their inputs are always raw `datetime` objects, as `Timestamp` objects should never use `dt.microsecond`.

> Ideally, as a side task, this check should be moved to the Timestamp constructor and unit tests for writer_binary_raw should be updated accordingly.

I agree, this bounds checking should be added to the constructor (and we should probably make note of this in the docstring once that's done). For example, a `ValueError` should be raised if the user tries to construct a `Timestamp` with `microsecond=999999` and `fractional_precision=3` because data would be lost.

Reply:
Unfortunately, I cannot add this bounds checking to the `Timestamp` constructor now. The same tests that previously complained a `ValueError` wasn't being thrown now complain that the `ValueError` weren't thrown within `writer_binary_raw`.
After this PR has been merged, bounds checking should be added to `Timestamp` and unit tests need to be updated.